tangelo/toolboxes/circuits/qsp.py

```diff
@@ -0,0 +1,272 @@
+        tau (float): The evolution time.
+        eps (float): The allowed error of the time-evolution. Higher accuracy requires longer circuits, and obtaining the phases can be unstable.
+        control (Union[int, List[int]]): The control qubit(s).
+        n_attempts (int): The number of attempts to calculate the phase factors using pyqsp.
+        method (str): "laurent": Use laurent polynomial method of pyqsp, "tf": Use TensorFlow with pyqsp, "QSPPACK": Use QSPPACK to calculate phases
+        folder (str): Folder that contains QSPPACK.
+
+    Returns:
+        Circuit: The QSP Hamiltonian simulation circuit with oblivious amplitude amplification to ensure very high success probability.
+    """
+
+    if control is not None:
+        control_list = copy(control) if isinstance(control, list) else [control]
+    else:
+        control_list = []
+
+    # If tau is negative, flip sign of tau and qu_op.
+    flip_tau_sign = False
+    if tau < 0.:
+        qu_op = -qu_op
+        tau = -tau
+        flip_tau_sign = True
+
+    qu_op_qs, m_qs, alpha = get_lcu_qubit_op_info(qu_op)
+
+    if method.lower() in ["laurent", "tf"]:
+        anglesr, anglesi = ham_sim_phases(tau*alpha, eps, n_attempts, method)
+    elif method.lower() == "qsppack":
+        anglesr, anglesi = ham_sim_phases_QSPPACK(folder, tau*alpha, eps)
+    else:
+        raise ValueError(f"{method} is not a valid keyword to calculate phases. Must be Laurent, TF, or QSPPACK")
+
+    # Want 1-norm of coefficents to sum to 1/np.sin(np.pi/(2*(2*3+1))) so three oblivious amplitude amplifications results
+    # in success probability of 1. |(cos(Ht)|=2 and |iSin(Ht))|=2 so need to add (tsum-4)/2 I - (tsum-4)/2 I.
+    tsum = 1/np.sin(np.pi/(2*(2*3+1)))
+    v = [(tsum-4)/2, (tsum-4)/2, 2, 2]
+    v = np.sqrt(np.array(v))/np.sqrt(tsum)
+
+    # Leave gap of 1-qubit for qsp_circuit and list LCU qubits to add cos(Ht) + Sin(Ht) + (tsum-4)/2 I - (tsum-4)/2 I
+    lcu_qs = list(range(m_qs[-1]+2, m_qs[-1]+4))
+    flip_qs = m_qs + [m_qs[-1]+1] + lcu_qs
+
+    uprep, uselect, qu_op_qs, m_qs, alpha = get_uprep_uselect(qu_op, control=lcu_qs+control_list)
+    cua = uprep + uselect + uprep.inverse()
+
+    s = StateVector(v, order="msq_first")
+    uprep = s.initializing_circuit()
+    uprep.reindex_qubits([lcu_qs[0], lcu_qs[1]])
+
+    circ = uprep + Circuit([Gate("X", lcu_qs[0])])
+    # real part cos(Ht)
+    circ += get_qsp_circuit_no_anc(cua, m_qs, anglesr, control=lcu_qs+control_list, with_z=False)
+    circ += Circuit([Gate("X", lcu_qs[0])])
+    # imaginary part i*sin(Ht)
+    circ += get_qsp_circuit_no_anc(cua, m_qs, anglesi, control=lcu_qs+control_list, with_z=False)
+    # -I to ensure probability is np.arcsin
+    circ += Circuit([Gate("X", lcu_qs[1]), Gate("CRZ", target=0, parameter=2*np.pi, control=lcu_qs+control_list), Gate("X", lcu_qs[1])])
+
+    circ += uprep.inverse()
+
+    if control is not None:
```

Comment:
Out of curiosity: do we need the branching ?
What if we only use the first branch, and control=None is passed ?

It seems to make sense to me that control=None should naturally mean it's not a controlled gate and would return something no different than the non-controlled version of the gate, as expected.

Reply:
Yes, in general that is correct. Our current implementation forces users to specify controlled gates with a `C` in front of the gate name.  We can certainly revisit that if you want.

Reply:
We can revisit that later if a user deems it a concern.